README.md

```diff
@@ -327,30 +315,178 @@ $(promote [d|
+  |])
+```
+
+A naïve attempt to promote `map` would be:
+
+```hs
+type Map :: (a -> b) -> [a] -> [b]
+type family Map f xs where
+  Map _ '[]    = '[]
+  Map f (x:xs) = f x : Map f xs
+```
+
+While this compiles, it is much less useful than we would like. In particular,
+common idioms like `Map Id xs` will not typecheck, since GHC requires that all
+invocations of type families be fully saturated. That is, the use of `Id` in
+`Map Id xs` is rejected since it is not applied to one argument, which the
+number of arguments that `Id` was defined with. For more information on this
+point, refer to the promotion paper.
+
+Not having the ability to partially apply functions at the type level is rather
+painful, so we do the next best thing: we _defunctionalize_ all promoted
+functions so that we can emulate partial application. For example, if one were
+to promote the `id` function:
+
+```hs
+$(promote [d|
+  id :: a -> a
+  id x = x
+  |]
+```
+
+Then in addition to generating the promoted `Id` type family, two
+defunctionalization symbols will be generated:
+
+```hs
+type IdSym0 :: a ~> a
+type IdSym0 x = x
+
+type IdSym1 (x :: a) = Id a
+```
+
+In general, a function that accepts _N_ arguments generates _N_ defunctionalization
+symbols when promoted.
+
+`IdSym1` is a _fully saturated_ defunctionalization symbol and is usually only
+needed when generating code through the Template Haskell machinery. `IdSym0`
+is more interesting: it has the kind `a ~> a`, which has a special arrow type
+`(~>)`. Defunctionalization symbols using the `(~>)` kind are type-level
+constants that can be "applied" using a special `Apply` type family:
+
+```hs
+type Apply :: (a ~> b) -> a -> b
+type family Apply f x
+```
+
+Every defunctionalization symbol comes with a corresponding `Apply` instance
+(except for fully saturated defunctionalization symbols). For instance, here
+is the `Apply` instance for `IdSym0`:
+
+```hs
+type instance Apply IdSym0 x = x
+```
+
+The `(~>)` kind is used when promoting higher-order functions so that partially
+applied arguments can be passed to them. For instance, here is a better attempt
+at promoting `map`:
+
+```hs
+type Map :: (a ~> b) -> [a] -> [b]
+type family Map f xs where
+  Map _ '[]    = '[]
+  Map f (x:xs) = Apply f x : Map f xs
+```
+
+Now `map id` can be promoted to `Map IdSym0`, which typechecks without issue.
+
+## Defunctionalizing existing type families
+
+The most common way to defunctionalize functions is by promoting them with the
+Template Haskell machinery. One can also defunctionalize existing type families,
+however, by using `genDefunSymbols`. For example:
+
+```hs
+type MyTypeFamily :: Nat -> Bool
+type family MyTypeFamily n
+
+$(genDefunSymbols [''MyTypeFamily])
+```
+
+This can be especially useful if `MyTypeFamily` needs to be implemented by
+hand. Be aware of the following design limitations of `genDefunSymbols`:
+
+* `genDefunSymbols` only works for type-level declarations. Namely, it only
```

Comment:
I thought we were going to have a `genAllDefunSymbols` or some such that went deeper...